### DISCUSSION TITLE
Rakefile Automation 1

### DISCUSSION THREAD

**Discussion by tajmone** [alan-if/alan-i18n#27] [Tech Stuff]
@thoni56, I've managed to automate in the `Rakefile` generating tasks for adventures with solution files. Now, to add a test folder you only need to:

```rake
    desc "English Test Suite"
    task :tests
    transcripts_folder_to_task(:tests,'alan_en/tests', LIB_EN_SOURCES)
```

See:

- https://github.com/alan-if/alan-i18n/blob/dev_alan-en/Rakefile

The custom method `transcripts_folder_to_task()` which I've created takes the following parameters:

* `target_task` -- The target task for storyfiles and transcripts.
* `target_folder` -- The folder that needs to be scanned.
* `dependencies` -- Required ALAN sources to build the storyfiles (e.g. a library).

The method is also smart enough to check if the folder contains just a single adventure or more than one, in the former case it associates all solutions and transcripts to the only adventure found, using the pattern `*.a3s`, in the latter case it uses the `<advname>*.a3s` pattern and associate dependencies accordingly.

Obviously, this makes Rakefiles creation and management much easier, and we don't have to edit the `Rakefile` each time a new test adventure or solution file is added (the method will handle it all by itself).

The next step will be to create a method to handle transcripts generation via ARun. BTW I've come up with a good solution for the problem of renaming transcripts like solutions without having to do too many file operations: I'll write a Ruby method that pipes the solution file to ARun, stripping off the BOM (if present), so that I we can safely use redirection of ARun output without the BOM ending up being injected into the final transcript (see alan-if/alan#32).

Now, I need some help with finding good names for these methods — and, _yes_, I did ready the whole _[Naming as a Process]_ series of articles! but I still can't come up with decent names. :cry:

The current name of the method `transcripts_folder_to_task()` doesn't seem all that good. Here's what this method does:

- Given a __target task__, __folder__ and __Alan dependencies__ (i.e. alan sources required to compile any adventure in the target folder) it will:
    + Scan the __target folder__ for any `*.alan` sources:
        * add `<advname>.a3c` as a dependency of the __target task__
        * create file task `<advname>.a3c => <advname>.alan`
        * add __Alan dependencies__ to `<advname>.a3c` file task
    + Scan the __target folder__ for any `*.a3s` sources:
        * add `<solname>.a3t` as a dependency of the __target task__
        * create file task `<solname>.a3t => [<solname>.a3s, <advname>.a3c]`

This method is intended to use whenever there's a folder with a sample adventure or test files, and it assumes that the folder also contains solution files — the real goal (target) are the transcript files, not compiling the adventures, for usually the transcripts are used to track any changes in the adventures or their dependencies (especially when dealing with libraries).

Ideally, I'd like to move this method to a Rake module, so it can be used with other projects too, and there will be more such methods coming soon.

Also consider that the method could be further improved, e.g. by making both the presence of solution files and the `dependencies` parameter optional, so that it might also be used just to compile Alan adventures in a folder — i.e. avoiding having to create another method to do that, since it could easily be done with this method too.

So, any suggestions what would be a good name for this method?

Also, any feedback and suggestions on the Rakefile progress so far?

## Method Source Code

<details>
<summary>(click to expand)</summary>

```rake
def transcripts_folder_to_task(target_task, target_folder, dependencies)
  # -----------------------------------------------------------------------
  # Process the target folder adding to the target task all the storyfiles
  # and transcripts as dependencies, and create all the required file tasks
  # dependencies. Can handle multiple adventures in a same folder, in which
  # case solutions will be associate to each adventure according to its
  # base filename: "<advname>*.sol"
  # -----------------------------------------------------------------------
  alan_sources = FileList["#{target_folder}/*.alan"]
  alan_sources.each do |alan_src|
    storyfile = alan_src.ext('.a3c')
    task target_task => storyfile
    file storyfile => alan_src
    file storyfile => dependencies
    if alan_sources.count() == 1
      # Single adventure: all solutions apply to it.
      solutions = FileList["#{target_folder}/*.a3s"]
    else
      # There are multiple adventures in a same folder!
      # Associate solutions using "<advname>*.sol" pattern.
      basename = storyfile.pathmap("%n")
      solutions = FileList["#{target_folder}/#{basename}*.a3s"]
    end
    solutions.each do |solution|
      transcript = solution.ext('.a3t')
      task target_task => transcript
      file transcript => [solution, storyfile] do |transcript|
        a3t = transcript.name.pathmap("%f")
        a3s = solution.pathmap("%f")
        a3c = storyfile.pathmap("%f")
        a3t_raw = a3c.ext('.a3t')

        hstr = "## Generating Transcript: #{a3t}"
        puts "\n#{hstr}"
        puts '#' * hstr.length

        cd "#{$repo_root}/#{target_folder}"
        sh "arun -r -l #{a3c} < #{a3s} 1>#{$devnull}"
        unless a3t_raw == a3t
          mv(a3t_raw, a3t, force: true)
        end
        cd $repo_root, verbose: false
      end
    end
  end
end
```
</details>

[Naming as a Process]: https://www.digdeeproots.com/articles/on/naming-process/

**Comment by thoni56** [comment]
This looks really sweet! I've not done any Rake-ing despite my ambition to do so, also my Ruby has not really improved over the last weeks. But I really appreciate your progress!

> Now, I need some help with finding good names for these methods — and, _yes_, I did ready the whole _[Naming as a Process](https://www.digdeeproots.com/articles/on/naming-process/)_ series of articles! but I still can't come up with decent names. 😢

Knowing and understanding is not always the same as being able to apply ;-)

> The current name of the method `transcripts_folder_to_task()` doesn't seem all that good. Here's what this method does:
> 
> * Given a **target task**, **folder** and **Alan dependencies** (i.e. alan sources required to compile any adventure in the target folder) it will:
>   
>   * Scan the **target folder** for any `*.alan` sources:
>     
>     * add `<advname>.a3c` as a dependency of the **target task**
>     * create file task `<advname>.a3c => <advname>.alan`
>     * add **Alan dependencies** to `<advname>.a3c` file task
>   * Scan the **target folder** for any `*.a3s` sources:
>     
>     * add `<solname>.a3t` as a dependency of the **target task**
>     * create file task `<solname>.a3t => [<solname>.a3s, <advname>.a3c]`
> 

I think this is the problem for you (in this case, but we often all are), too close to what it does internally and not thinking enough about the external purpose. It looks like you are on level 5 ;-). I would say that levels 6 and 7, the "Intent revealing" to "Domain Abstraction", would give you a name like `create_transcripting_tasks_from_folder()` or something like that ("create" and "from folder" being intent revealing and "transcripting task" being the "domain abstraction"). Is that a good suggestion?

(I was thinking about `running.....in_folder()` but that is not the way Rake works, so I scrapped that.)

> Also consider that the method could be further improved, e.g. by making both the presence of solution files and the `dependencies` parameter optional, so that it might also be used just to compile Alan adventures in a folder — i.e. avoiding having to create another method to do that, since it could easily be done with this method too.

My advice would be to not put too much features in the same function, at first at least. A suggestion would be to duplicate the function, add the new features in the copy, make it work. clean it up and then refactor to extract the common parts. That usually make better and cleaner "parts" than doing it the other way. But of course, YMMV and context is king.

**Reply by tajmone** [reply]
> I would say that levels 6 and 7, the "Intent revealing" to "Domain Abstraction", would give you a name like `create_transcripting_tasks_from_folder()` or something like that ("create" and "from folder" being intent revealing and "transcripting task" being the "domain abstraction"). Is that a good suggestion?

Yes that sounds great! thank you.

> I think this is the problem for you (in this case, but we often all are), too close to what it does internally and not thinking enough about the external purpose

Yes, that's not easy. I find it difficult to balance between focusing on the Rake side, on the one hand, and the project on the other — I either end up giving granted the former and focus too much on the latter, or vice-versa.

> My advice would be to not put too much features in the same function, at first at least. A suggestion would be to duplicate the function, add the new features in the copy, make it work. clean it up and then refactor to extract the common parts. That usually make better and cleaner "parts" than doing it the other way. But of course, YMMV and context is king.

I thought of that, but haven't really got to work it out in Ruby. I'm still in my very early Ruby learning stages — I've read a whole beginner's book of 500 pages, but Ruby is quite a huge language, and still haven't managed to work out which is the best way to do anything idiomatically.

And, of course, there are also the future goals of providing some more ALAN helpers via custom Rake modules, which will probably become an ALAN class, with interfaces to the compiler, to ARun, and extra methods for the SDK (checking version, downloading and updating new SDKs in the project, etc.). Probably once I start implementing all this as a Ruby module, with classes and all, it will become clearer which way to go about.

I'm really liking Ruby, it's a very powerful language. It has tons of native features, so just learning those is going to be take quite a while, and then there's all the Ruby Gems ecosystem, which is also very rich indeed and demands being explored. I'm not used to work with duck typing, so it feels a bit strange right now, but that and everything being a class in Ruby, everything feels seamless. Ruby 3 brings quite a bunch of new cool features, both stable and experimental, and indeed makes Ruby _much_ faster than version 2 (up to three times faster in some cases); and I did notice that after switching to v3 all our Asciidoctor toolchains become faster.
